src/common/updating/updating.cpp

```diff
@@ -38,46 +72,48 @@ namespace updating
-            winrt::Windows::Foundation::Uri release_page_uri{ json_body.GetNamedString(L"html_url") };
-
-            VersionHelper github_version(winrt::to_string(new_version));
-            VersionHelper current_version(VERSION_MAJOR, VERSION_MINOR, VERSION_REVISION);
-
-            if (github_version <= current_version)
+            json::JsonObject release_object;
+            const VersionHelper current_version(VERSION_MAJOR, VERSION_MINOR, VERSION_REVISION);
+            VersionHelper github_version = current_version;
+            if (prerelease)
             {
-                co_return nonstd::make_unexpected(strings.GITHUB_NEW_VERSION_UP_TO_DATE);
-            }
-
-            const std::wstring_view required_architecture = get_architecture_string(get_current_architecture());
-            constexpr const std::wstring_view required_filename_pattern = updating::INSTALLER_FILENAME_PATTERN;
-            // Desc-sorted by its priority
-            const std::array<std::wstring_view, 2> asset_extensions = { L".exe", L".msi" };
-            for (const auto asset_extension : asset_extensions)
-            {
-                for (auto asset_elem : json_body.GetNamedArray(L"assets"))
+                const auto body = co_await client.request(Uri{ ALL_RELEASES_ENDPOINT });
+                for (const auto& json : json::JsonValue::Parse(body).GetArray())
                 {
-                    auto asset{ asset_elem.GetObjectW() };
-                    std::wstring filename_lower = asset.GetNamedString(L"name", {}).c_str();
-                    std::transform(begin(filename_lower), end(filename_lower), begin(filename_lower), ::towlower);
-
-                    const bool extension_matched = filename_lower.ends_with(asset_extension);
-                    const bool architecture_matched = filename_lower.find(required_architecture) != std::wstring::npos;
-                    const bool filename_matched = filename_lower.find(required_filename_pattern) != std::wstring::npos;
-                    if (extension_matched && architecture_matched && filename_matched)
+                    auto potential_release_object = json.GetObjectW();
+                    const bool is_prerelease = potential_release_object.GetNamedBoolean(L"prerelease", false);
+                    auto extracted_version = extract_version_from_release_object(potential_release_object);
+                    if (!is_prerelease || extracted_version <= github_version)
                     {
-                        winrt::Windows::Foundation::Uri msi_download_url{ asset.GetNamedString(L"browser_download_url") };
-                        co_return new_version_download_info{ std::move(release_page_uri), new_version.c_str(), std::move(msi_download_url), std::move(filename_lower) };
+                        continue;
                     }
+                    github_version = std::move(extracted_version);
+                    release_object = std::move(potential_release_object);
```

Comment:
`for` loop could be stopped after finding the release object.

Reply:
right, that's actually a good place to add a comment!
the reason I've done it this way is that https://developer.github.com/v3/repos/releases/#list-releases doesn't clearly specify the default sorting order of the releases objects (or did I miss it?), so I just did an additional precaution and assumed it could be changed w/o any warning.